### PR TITLE
NMS-13359: Improved Visibility into Tags

### PR DESCRIPTION
**Pull request by patrick-schweizer**
This PR adds measurements to the timeseries integration layer.

We measure:

- What metrics series have the highest tag cardinality?
  - What does the set tags for the top 10 look like?
- Which string properties have the most unique values?

* JIRA (Issue Tracker): https://issues.opennms.org/browse/NMS-13359

### CODE REVIEW

File: features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            topNMetrics.pollLast();
+        }
+        this.lowestTagCount.set(countNoOfTags(topNMetrics.last()));
+    }
+
+    void putInTopTags(final Tag tag) {
+        @SuppressWarnings("UnstableApiUsage")
+        long hash = hllHashFunction.hashString(tag.getValue(), StandardCharsets.UTF_8).asLong();
+        topNTags.computeIfAbsent( tag.getKey(), k -> new HLL(14, 5))
+                .addRaw(hash);
```
Comment (j-white) [inline features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java:92]: The HLL is not thread safe so we need to make sure that the call to `.addRaw` is synchronized here.

Would be good to use a striped lock for this, where the key the for lock is the key of the hash table.

File: features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    private final ConcurrentSkipListSet<Metric> topNMetrics = new ConcurrentSkipListSet<>(
+            Comparator.comparingInt(m -> ((Metric)m).getMetaTags().size() + ((Metric)m).getExternalTags().size())
+                    .reversed()
+                    .thenComparing(m -> ((Metric)m).getKey())); // write should not happen very often since we just push to the highest limit
+    private final ConcurrentHashMap<String, HLL> topNTags = new ConcurrentHashMap<>();
+    @SuppressWarnings("UnstableApiUsage")
+    private final HashFunction hllHashFunction = Hashing.murmur3_128();
+
+    public void record(Collection<Sample> samples) {
+        // race conditions might happen but it shouldn't matter for statistical purposes.
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java:65]: The method is not synchronized. Therefor the count of topNMetrics and lowestTagCount might not be always correct:
`if (topNMetrics.size() < MAX_TOP_N_METRIC && count >= lowestTagCount.get() || count > lowestTagCount.get() ) {`

File: features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java (hunk 1 above)
Comment (patrick-schweizer) [reply features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java]: I added the strip lock. Please check if ok. Thanks!

File: features/timeseries-shell/src/main/java/org/opennms/netmgt/timeseries/shell/TimeseriesStatsCommand.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        System.out.println("Tags with highest number of unique values (top 100):");
+        stats.getTopNTags().stream().limit(100).forEach(out::println);
+        return null;
+    }
+
+    private String toString(final Metric metric) {
+        return metric.getFirstTagByKey(IntrinsicTagNames.resourceId).getValue() + "/"
+                + metric.getFirstTagByKey(IntrinsicTagNames.name).getValue() +
+                "\n    metaTags:      " + metric.getMetaTags().toString() +
+                "\n    intrinsicTags: " + metric.getIntrinsicTags().toString();
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply features/timeseries-shell/src/main/java/org/opennms/netmgt/timeseries/shell/TimeseriesStatsCommand.java]: done. They should have been here in the first place :-/

File: features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (j-white) [reply features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java:79]: got it. can we elaborate on this in the comment?

**Comment by j-white** [comment]
Looks good!

**Review by j-white — APPROVED** [review]

File: features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply features/timeseries/src/main/java/org/opennms/netmgt/timeseries/stats/StatisticsCollectorImpl.java]: done.

**Pull request merged**